implementations/rust/ockam/ockam/src/credential.rs

```diff
@@ -1,44 +1,270 @@
+#![allow(missing_docs)]
+
 //! Building block for Attribute-based access control
 
 mod exchange;
 
-pub use exchange::*;
-
+use core::time::Duration;
+use minicbor::bytes::ByteSlice;
 use minicbor::{Decode, Encode};
-use ockam_core::CowBytes;
+use ockam_core::compat::borrow::Cow;
+use ockam_core::compat::collections::BTreeMap;
+use ockam_core::errcode::{Kind, Origin};
+use ockam_core::vault::{Signature, SignatureVec};
+use ockam_core::{CowBytes, CowStr, Error, Result};
+use ockam_identity::{Identity, IdentityIdentifier, IdentityVault};
+
+pub use exchange::*;
 
 #[cfg(feature = "tag")]
 use crate::TypeTag;
 
-/// Credential
+pub const MAX_CREDENTIAL_VALIDITY: Duration = Duration::from_secs(6 * 3600);
+
 #[derive(Debug, Decode, Encode)]
 #[rustfmt::skip]
 #[cbor(map)]
 pub struct Credential<'a> {
     #[cfg(feature = "tag")]
     #[n(0)] tag: TypeTag<3796735>,
+    /// CBOR-encoded [`CredentialData`].
     #[b(1)] attributes: CowBytes<'a>,
+    /// Cryptographic signature of attributes data.
     #[b(2)] signature: CowBytes<'a>,
 }
 
+#[derive(Debug, Decode, Encode)]
+#[rustfmt::skip]
+#[cbor(map)]
+pub struct CredentialData<'a> {
+    /// A schema identifier to allow distinguishing sets of attributes.
+    #[n(1)] schema: Option<SchemaId>,
+    /// User-defined key-value pairs.
+    #[b(2)] attributes: Option<Attributes<'a>>,
```

Comment:
Not sure it makes sense to make it Optional. Usually it will have some values (since that the whole point of Credential), otherwise it can also be empty

Reply:
I used an `Option` as an easy easy way to not encode the key and the empty map, but this can be postponed (if needed at all).